src/relay/pass/mac_count.cc

```diff
@@ -0,0 +1,138 @@
+/*!
+ * Copyright (c) 2019 by Contributors
+ *
+ * \file mac_count.cc
+ * \brief Pass to roughly count the number of MACs (Multiply-Accumulate) 
+ * operations of a model. Only MACs in CONV and Dense ops are counted.
+ * This pass is valid after the type infer pass is called,
+ * otherwise the count is 0.
+ */
+
+#include <tvm/relay/op.h>
+#include <tvm/relay/attrs/nn.h>
+#include <tvm/relay/expr_functor.h>
+
+namespace tvm {
+namespace relay {
+
+namespace {
+
+bool IsConv2DNode(const ExprNode* node) {
+  const auto* call_node = dynamic_cast<const CallNode*>(node);
+  return call_node != nullptr && call_node->attrs.as<Conv2DAttrs>();
+}
+
+bool IsDenseNode(const ExprNode* node) {
+  const auto* call_node = dynamic_cast<const CallNode*>(node);
+  return call_node != nullptr && call_node->attrs.as<DenseAttrs>();
+}
+
+}  // namespace
+
+class MacCounter : private ExprVisitor {
+ public:
+  MacCounter() {
+      count_ = 0;
+  }
+  static int64_t GetTotalMacNumber(const Expr& expr) {
+    MacCounter counter;
+    counter(expr);
+    return counter.count_;
+  }
+
+ private:
+  void VisitExpr_(const CallNode* call_node) final {
+    if (IsConv2DNode(call_node)) {
+      count_ += ComputeConv2DMacs(call_node);
+    } else if (IsDenseNode(call_node)) {
+      count_ += ComputeDenseMacs(call_node);
+    }
+    ExprVisitor::VisitExpr_(call_node);
+  }
+
+  /*
+   * \brief Get the number of MACs of a CONV 2D node.
+   * \param call_node The CONV 2D call node.
+   * \return The number of MACs.
+   */
+  int64_t ComputeConv2DMacs(const CallNode* call_node) {
+    CHECK(IsConv2DNode(call_node))
+        << "The input call node must be a CONV 2D node.";
+    if (!call_node->checked_type_.defined()) {
+      LOG(WARNING) << "The infer type pass should be called before the mac count pass";
+      return 0;
+    }
+    Array<Expr> args = call_node->args;
+    CHECK(args.size() == 2)
+        << "The number of input arguments of a CONV 2D node should be 2.";
+    const auto* data_type = args[0]->checked_type().as<TensorTypeNode>();
+    Array<IndexExpr> data_shape = data_type->shape;
+    CHECK(data_shape.size() == 4)
+        << "The dimension of an input tensor to Dense node should be 4.";
```

Comment:
```suggestion
        << "The dimension of an input tensor to Conv2D node should be 4.";
```
some non-4D conv implement also use Conv2dAttrs though... I think the dimension of input & output does not really matter.

Reply:
If a non-4D conv uses Conv2dAttrs, perhaps we should pick it out, otherwise the output wouldn't make sense?